pkg/sdk/send_relay.go

```diff
@@ -0,0 +1,109 @@
+package sdk
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log"
+	"net/http"
+
+	cryptocomet "github.com/cometbft/cometbft/crypto"
+	"github.com/pokt-network/poktroll/pkg/signer"
+	"github.com/pokt-network/poktroll/x/service/types"
+)
+
+// SendRelay sends a relay request to the given supplier's endpoint.
+// It signs the request, relays it to the supplier and verifies the response signature.
+// It takes an http.Request as an argument and uses its method and headers to create
+// the relay request.
+func (sdk *poktrollSDK) SendRelay(
+	ctx context.Context,
+	supplierEndpoint *SupplierEndpoint,
+	request *http.Request,
+) (response *types.RelayResponse, err error) {
+	payloadBz, err := io.ReadAll(request.Body)
+	if err != nil {
+		return nil, ErrSDKHandleRelay.Wrapf("reading request body: %s", err)
+	}
+
+	// Create the relay request.
+	relayRequest := &types.RelayRequest{
+		Meta: &types.RelayRequestMetadata{
+			SessionHeader: supplierEndpoint.Header,
+			Signature:     nil, // signature added below
+		},
+		Payload: payloadBz,
+	}
+
+	// Get the application's signer.
+	appAddress := supplierEndpoint.Header.ApplicationAddress
+	appRing, err := sdk.ringCache.GetRingForAddress(ctx, appAddress)
+	if err != nil {
+		return nil, ErrSDKHandleRelay.Wrapf("getting app ring: %s", err)
+	}
+	signer := signer.NewRingSigner(appRing, sdk.signingKey)
+
+	// Hash and sign the request's signable bytes.
+	signableBz, err := relayRequest.GetSignableBytes()
+	if err != nil {
+		return nil, ErrSDKHandleRelay.Wrapf("getting signable bytes: %s", err)
+	}
+
+	hash := cryptocomet.Sha256(signableBz)
```

Comment:
This seems odd. We should be signing EXACTLY what `GetSignableBytes` returns.

Reply:
We should then change `GetSignableBytes` implementation since it's not hashing the marshalled structure.

Previously answered here https://github.com/pokt-network/poktroll/pull/199#discussion_r1414325367